x-pack/elastic-agent/pkg/composable/providers/kubernetes/kubernetes.go

```diff
@@ -130,19 +137,19 @@ func (p *dynamicProvider) newWatcher(
 	config *Config) (kubernetes.Watcher, error) {
 	switch resourceType {
 	case "pod":
-		watcher, err := NewPodWatcher(comm, config, p.logger, client, p.config.Scope)
+		watcher, err := NewPodWatcher(comm, config, p.logger, client, p.config.Scope, p.rawConfig)
```

Comment:
It looks a bit weird to me to need to pass another config, raw, to this method. Some thoughts about this, no need to do anything now.

The `Config` type of `config` is quite similar to the `metadata.Config` type used by the underlying metadata generator. I guess that the new Config object was introduced here to decouple the composable provider config, but at this point they are becoming more similar, I wonder if they could became the same thing.

Talking about decoupling, we are implementing another set of watchers here, would it make sense to also implement a new set of metadata generators here? I am totally +1 with reusing them, but I wonder if this would be a chance to clean something.

Other thing that could be explored could be to refactor `metadata` package, so the constructors there receive a `metadata.Config` instead of a raw `common.Config`. This would allow to more cleanly reuse them, for example here it would allow to generate or unpack this metadata config here, and pass it to the watchers, instead of the raw config.

Reply:
I cahnge the code a little bit in order to avoid moving around the configs (both basic and raw). 


As far as the decoupling is concerned I think that it's not such a good idea since meta Generators are already used by several other parts like `add_k8s_metadata`, k8s module and k8s autodiscover (beats) so it stands as a single source of truth for metadata. Since the porcessor and the module are coming along to Agent's era it would be nice to preserve this common ground for them and have a single metadata generation library. Just my personal point of view but so far it has been quite useful to handle meta from a single place, happy to discuss it more though if there is a better approach. 


Regarding the cleanups/reorganisation of generators I'm totally +1 but I would wait to see how the projects will evolve in the near future and work on cleanups while we move (I still have hopes that hints are coming soon :)).